operators/src/pro/cache/shared_cache.rs

```diff
@@ -0,0 +1,1190 @@
+
+    cache_size: CacheSize,
+    landing_zone_size: CacheSize,
+
+    // we only use the LruCache for determining the least recently used elements and evict as many entries as needed to fit the new one
+    lru: LruCache<CacheEntryId, TypedCanonicOperatorName>,
+}
+
+impl CacheBackend {
+    /// This method removes entries from the cache until it can fit the given amount of bytes.
+    pub fn evict_until_can_fit_bytes(&mut self, bytes: usize) {
+        while !self.cache_size.can_fit_bytes(bytes) {
+            if let Some((pop_id, pop_key)) = self.lru.pop_lru() {
+                match pop_key {
+                    TypedCanonicOperatorName::Raster(raster_pop_key) => {
+                        let op_cache = self
+                            .raster_caches
+                            .get_mut(&raster_pop_key)
+                            .expect("LRU entry must exist in the cache!");
+                        let query_element = op_cache
+                            .remove_cache_entry(&pop_id)
+                            .expect("LRU entry must exist in the cache!");
+                        self.cache_size.remove_element_bytes(&query_element);
+                    }
+                    TypedCanonicOperatorName::Vector(vector_pop_key) => {
+                        let op_cache = self
+                            .vector_caches
+                            .get_mut(&vector_pop_key)
+                            .expect("LRU entry must exist in the cache!");
+                        let query_element = op_cache
+                            .remove_cache_entry(&pop_id)
+                            .expect("LRU entry must exist in the cache!");
+                        self.cache_size.remove_element_bytes(&query_element);
+                    }
+                };
+                self.cache_size.remove_element_bytes(&pop_id);
+
+                debug!(
+                    "Evicted query {}. Cache size: {}. Cache size used: {}, Cache used percentage: {}.",
+                    pop_id,
+                    self.cache_size.total_byte_size(),
+                    self.cache_size.byte_size_used(),
+                    self.cache_size.size_used_fraction()
+                );
+            }
+        }
+    }
+}
+
+#[derive(Clone, Debug, PartialEq, Eq, Hash)]
+pub enum TypedCanonicOperatorName {
+    Raster(CanonicOperatorName),
+    Vector(CanonicOperatorName),
+}
+
+impl TypedCanonicOperatorName {
+    pub fn as_raster(&self) -> Option<&CanonicOperatorName> {
+        match self {
+            Self::Raster(name) => Some(name),
+            Self::Vector(_) => None,
+        }
+    }
+
+    pub fn as_vector(&self) -> Option<&CanonicOperatorName> {
+        match self {
+            Self::Raster(_) => None,
+            Self::Vector(name) => Some(name),
+        }
+    }
+}
+
+pub trait CacheEvictUntilFit {
+    fn evict_entries_until_can_fit_bytes(&mut self, bytes: usize);
+}
+
+impl CacheEvictUntilFit for CacheBackend {
```

Comment:
why don't you implement this directly on CacheBackend. What is the trait useful for?

Reply:
the trait is needed to "derive" the other trait from. This way i can access the method in the trait methods.